llvm/utils/repo/Makefile

```diff
@@ -0,0 +1,121 @@
+TOOLCHAIN_FILE=$(LLVM_REPO)/llvm/utils/repo/repo.cmake
+# The path of the llvm-project-prepo build directory.
+LLVM_BUILD = ${LLVM_REPO}/build
+# The path of the directory in which llvm libraries are installed.
+LLVM = ${HOME}/LLVM
+# The path of the directory in which musl libc is installed.
+MUSL = ${HOME}/musl
+
+libunwind.a:
+	if [ -d "$(LLVM_REPO)/build_libunwind" ];then                         \
+		rm -rf $(LLVM_REPO)/build_libunwind;                          \
+	fi
+	cd $(LLVM_REPO)  &&                                                   \
+	mkdir build_libunwind && cd build_libunwind  &&                       \
+	cmake -D musl=Yes                                                     \
+		-D CMAKE_BUILD_TYPE=Release                                   \
+		-D LLVM_PATH=../                                              \
+		-D LIBUNWIND_ENABLE_SHARED=No                                 \
+		-D LIBUNWIND_ENABLE_STATIC=Yes                                \
+		-D CMAKE_TOOLCHAIN_FILE=$(TOOLCHAIN_FILE)                     \
+		-D CMAKE_INSTALL_PREFIX=$(LLVM)                               \
+		../libunwind &&                                               \
+	make -j 8 &&                                                          \
+	make install
+
+libcompiler-rt.a:
+	if [ -d "$(LLVM_REPO)/build_libcompiler-rt" ];then                    \
+		rm -rf $(LLVM_REPO)/build_libcompiler-rt;                     \
+	fi
+	cd $(LLVM_REPO)  &&                                                   \
+	mkdir build_libcompiler-rt && cd build_libcompiler-rt  &&             \
+	cmake  -D musl=Yes                                                    \
+		-D CMAKE_BUILD_TYPE=Release                                   \
+		-D LLVM_CONFIG_PATH=$(LLVM_BUILD)/bin/llvm-config             \
+		-D COMPILER_RT_DEFAULT_TARGET_ONLY=ON                         \
+		-D COMPILER_RT_BUILD_BUILTINS=ON                              \
+		-D COMPILER_RT_BUILD_SANITIZERS=OFF                           \
+		-D COMPILER_RT_BUILD_XRAY=OFF                                 \
+		-D COMPILER_RT_BUILD_LIBFUZZER=OFF                            \
+		-D COMPILER_RT_BUILD_PROFILE=OFF                              \
+		-D CMAKE_TOOLCHAIN_FILE=$(TOOLCHAIN_FILE)                     \
+		-D CMAKE_INSTALL_PREFIX=$(LLVM)                               \
+		../compiler-rt &&                                             \
+	make -j 8 &&                                                          \
+	make install
+
+$(LLVM)/lib/linux/clang_rt.crtbegin-x86_64.o: libcompiler-rt.a
+$(LLVM)/lib/linux/clang_rt.crtend-x86_64.o: libcompiler-rt.a	
+
+$(LLVM)/lib/linux/clang_rt.crtbegin-x86_64.o.elf: $(LLVM)/lib/linux/clang_rt.crtbegin-x86_64.o
+	repo2obj --repo=$(LLVM_REPO)/build_libcompiler-rt/lib/crt/clang.db $< -o $@
+
+$(LLVM)/lib/linux/clang_rt.crtend-x86_64.o.elf: $(LLVM)/lib/linux/clang_rt.crtend-x86_64.o
+	        repo2obj --repo=$(LLVM_REPO)/build_libcompiler-rt/lib/crt/clang.db $< -o $@

```

Comment:
Why do we convert crtbegin and crtend to ELF here?

The second rule seems to have extra whitespace at the start.

Reply:
> Please update the comment to explain this a little more clearly.

Added in the commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/9400f67b65927921e3f5134453335a67197d8fc9.

Reply:
>
> The second rule seems to have extra whitespace at the start.
>

Thanks, fixed in the commit https://github.com/SNSystems/llvm-project-prepo/commit/7ef52d812cc1e524350d3241e5bdf90077da4b3e.

Reply:
> 
> 
> Why do we convert crtbegin and crtend to ELF here?
> 

When building the compiler-rt library, the crtbegin and crtend are not included in the libcompiler-rt.a library. The repo link (using link.py) and ar (using archive.py) involve to use the repo2obj to generate the ELF file. Since the crtbegin and crtend are only compiled, no corresponding crtbegin and crtend ELF files are generated. Therefore, crtbegin and crtend are converted ELF files here.